scipy/stats/_multivariate.py

```diff
@@ -510,12 +533,13 @@ def logpdf(self, x, mean=None, cov=1, allow_singular=False):
         %(_mvn_doc_callparams_note)s
 
         """
-        dim, mean, cov = self._process_parameters(None, mean, cov)
+        dim, mean, cov_object = self._process_parameters(mean, cov,
+                                                         allow_singular)
         x = self._process_quantiles(x, dim)
-        psd = _PSD(cov, allow_singular=allow_singular)
-        out = self._logpdf(x, mean, psd.U, psd.log_pdet, psd.rank)
-        if allow_singular and (psd.rank < dim):
-            out_of_bounds = ~psd._support_mask(x-mean)
+        out = self._logpdf(x, mean, cov_object)
+        allow_singular = allow_singular or cov_object._allow_singular
+        if allow_singular and np.any(cov_object.rank < dim):
```

Comment:
> doesn't `rank < dim` imply `allow_singular=True`

No. Unless the user explicitly passes `allow_singular=True`, `multivariate_normal` methods error out if the covariance is singular. I don't agree with this choice, but it's what it is. I'm choosing to ignore `allow_singular` when a new `Covariance` object is passed because they give the user much greater control about how to handle near-zero eigenvalues anyway.

Reply:
I didn't notice that the `allow_singular` check was not necessary in gh-5288. That was in the original PR, and I did not remove it. I also didn't question it here. But you're right, it appears to be unnecessary, so I've removed that part.

I haven't gone further toward refactoring. I think that can be done as a follow-up. I hesitate because I'm not sure exactly what the right thing to do is. For one, there's a question of what the name of the method should be. I don't think `mask_invalid` is appropriate because it suggests this has something to do with masked arrays. (Perhaps `_support_mask` is not perfect, but it communicates the idea that the "mask", which is obviously a normal boolean array, has to do with whether the argument is within the *support* of the distribution.) But also, there's a question of what `whiten` should be doing here. What is the meaning of whitening a point that is not in the subspace spanned by the columns of the covariance matrix? So the question is whether `whiten` should just do something that would obviate the need for a separate method. For now, I'll just leave it alone, but this is a good thing to consider for a follow-up.